Feature/export workflow

adoLime opened the pull request:
Task 1 of proposal done. Implemented a functional export workflow button to the timeline tab.

Commit:
Added the export button to the timeline

MxLinh committed:
added condition for export workflow button

MxLinh committed:
Revert "added condition for export workflow button"

This reverts commit b01525d9c5ad11e074e79f9fdcff2b189553dbc7.

MxLinh committed:
Added condition and click for export workflow button

2nd commit, since I've accidententally committed with the wrong git account :)

douirili committed:
Fixed:  “workflowGroupExists” and “exportWorkflow” now exist

The property and method are now exist in the corresponding file

MxLinh committed:
Added defensive programming and a helper method

Commit:
The methode getTemplateIdForExperiment is updated

emyyoussef2006 committed:
Rename of the identifier

MxLinh committed:
Implemented two helper methods

douirili committed:
Implemented 'exportWorkflow' method

Commit:
Fixed url

douirili committed:
Implemented 'buildBpmnXml' method

Commit:
fixed naming failure in bpmn by subscribing to experimentName and adding a corresponding attribute to the component. Also fixed the formatting

douirili committed:
Removed unused variable 'activeTabId'

Commit:
Add comments

buehlefs commented on src/app/components/experiment-timeline/experiment-timeline.component.ts:
```diff
@@ -132,4 +170,291 @@ export class ExperimentTimelineComponent implements OnInit, OnDestroy {
+                    console.log(steps);
+                    const xml = this.buildBpmnXml(steps);
+
+                    const postUrl = `http://localhost:5005/plugins/workflow-editor@v0-1-0/workflows/`;
```
this URL should not be hardcoded. You should be able to get this URL from the plugin metadata of the workflow plugin. To get to the plugin metadata, use the plugin registry service.

adoLime replied:
We are not sure how to get this URL dynamically.

MxLinh replied:
> We are not sure how to get this URL dynamically.

This has been fixed now. We got the URL dynamically.

Commit:
Extraction of processTimelineSteps() from exportWorkflow() method

Commit:
Get currentTemplateId via template-service

Commit:
Refactoring of processTimelineSteps to use currentTemplateId

Commit:
Removed getTemplateIdForExperiment-Method due to the fact we have a component variable for the currentTemplateId. Therefore the checkWorkflowGroup was also refactored to use it

MxLinh committed:
Generated a new component export-workflow-modal

Co-Authored-By: Lina Douiri <212174522+douirili@users.noreply.github.com>
Co-Authored-By: adoLime <232741651+adoLime@users.noreply.github.com>

MxLinh committed:
Fully implemented the feature to filter by quality for export workflow

Co-Authored-By: Lina Douiri <212174522+douirili@users.noreply.github.com>
Co-Authored-By: adoLime <232741651+adoLime@users.noreply.github.com>

Commit:
Refactoring of exportWorkflow-method to use the processTimelineSteps method again.
Also applying the latest changes from the exportWorkflow-method to the processTimelineSteps-method

Co-Authored-By: Lina Douiri <212174522+douirili@users.noreply.github.com>
Co-Authored-By: Mai Linh Dang <linhdangmai0107@gmail.com>

douirili committed:
Fix findings from PR #90

- refactoring: add 'bpmn-xml-builder'
- formatting: add .vscode/settings.json
- adapt experiment-timeline.component.ts accordingly

MxLinh committed:
fixed: resolve workflow editor URL dynamically

Co-Authored-By: adoLime <232741651+adoLime@users.noreply.github.com>
Co-Authored-By: Lina Douiri <212174522+douirili@users.noreply.github.com>

Commit:
The workflow editor version also changes dynamically now

MxLinh committed:
every query parameter will be passed to the micro frontend now

Co-Authored-By: adoLime <232741651+adoLime@users.noreply.github.com>
Co-Authored-By: Lina Douiri <212174522+douirili@users.noreply.github.com>

Commit:
Add Camunda input/output parameters to BPMN export. The workflow is now executable as plugin.

Co-Authored-By: Lina Douiri <212174522+douirili@users.noreply.github.com>
Co-Authored-By: Mai Linh Dang <linhdangmai0107@gmail.com>

Commit:
Parameter value "***" will be replaced with an empty string.

Co-Authored-By: Lina Douiri <212174522+douirili@users.noreply.github.com>
Co-Authored-By: Mai Linh Dang <linhdangmai0107@gmail.com>

MxLinh committed:
added condition to export button: show only if workflowExists and stepsExist

Added a new condition that needs to be met alongside with workflowExists. Only if there is at least one timeline step and the workflow tab exists, the button is displayed.

Co-Authored-By: adoLime <232741651+adoLime@users.noreply.github.com>

MxLinh committed:
set start value of stepsExist back to false

forgot the reset

Co-Authored-By: adoLime <232741651+adoLime@users.noreply.github.com>

buehlefs commented on src/app/components/experiment-timeline/bpmn-xml-builder.ts:
```diff
@@ -0,0 +1,359 @@
+    processorName: string;
+    processorVersion: string;
+    notes: string;
+    inputData: Array<{ type: string; contentType: string; name: string; version: number }>;
```
Could be extracted as their own type definitions.

buehlefs commented on src/app/components/experiment-timeline/bpmn-xml-builder.ts:
```diff
@@ -0,0 +1,359 @@
+    }
+
+    public toString(): string {
+        const processOpen = `<bpmn2:process id="Experiment_${this.experimentName}" isExecutable="true" name="Experiment_${this.experimentName}" camunda:historyTimeToLive="360000">`;
```
Process id should preferably not contain spaces

buehlefs commented on src/app/components/experiment-timeline/bpmn-xml-builder.ts:
```diff
@@ -0,0 +1,359 @@
+    private stepInputMappings: InputMapping[][] = [];
+
+    public constructor(
+        private experimentName: any,
```
why not `string` instead of `any`?

buehlefs commented on src/app/components/experiment-timeline/experiment-timeline.component.ts:
```diff
@@ -132,4 +172,173 @@ export class ExperimentTimelineComponent implements OnInit, OnDestroy {
+        this.backend
+            .getTimelineStepsPage(this.experimentId, {
+                page: 0,
+                itemCount: 100,
```
This will only handle experiments with <= 100 steps.

buehlefs commented on src/app/components/experiment-timeline/experiment-timeline.component.ts:
```diff
@@ -132,4 +172,173 @@ export class ExperimentTimelineComponent implements OnInit, OnDestroy {
+    private getWorkflowEditorHref(templateId: string): Observable<string> {
+        return this.getWorkflowTab(templateId).pipe(
+            switchMap((tabId) => {
+                const url = `${this.registry.registryRootUrl}plugins/?template-tab=${tabId}`;
```
preferably this code should use the registry client directly and not build URLs like this.

buehlefs commented on src/app/components/experiment-timeline/experiment-timeline.component.ts:
```diff
@@ -132,4 +172,173 @@ export class ExperimentTimelineComponent implements OnInit, OnDestroy {
+                const enrichedSteps = await this.fetchEnrichedSteps(selectedSteps);
+                const xml = new BpmnXmlBuilder(this.experimentName, enrichedSteps).toString();
+
+                this.getWorkflowEditorHref(this.currentTemplateId!).subscribe(href => {
```
TODO: Make sure that subscription gets unsubscribed.

buehlefs commented on src/app/components/experiment-timeline/experiment-timeline.component.ts:
```diff
@@ -132,4 +172,173 @@ export class ExperimentTimelineComponent implements OnInit, OnDestroy {
+    }
+
+    private navigateToTabId(tabId: string): void {
+        console.log(`Switching to workflow tab: ${tabId}`);
```
TODO: Remove debug outputs

buehlefs commented on src/app/components/experiment-timeline/experiment-timeline.component.ts:
```diff
@@ -132,4 +172,173 @@ export class ExperimentTimelineComponent implements OnInit, OnDestroy {
+        const queryParams = Object.fromEntries(
+            Object.entries(currentQueryParams)
+                .filter(([key]) => key.startsWith('param-'))
+                .map(([key, value]) => [key.replace(/^param-/, ''), value])
```
Why remove the `param-` prefix here?

buehlefs commented on src/app/components/experiment-timeline/experiment-timeline.component.ts:
```diff
@@ -132,4 +172,173 @@ export class ExperimentTimelineComponent implements OnInit, OnDestroy {
+     * and updates the workflowExists property accordingly
+     */
+    private checkWorkflowGroup(experimentId: string): void {
+        const url = `${this.registry.registryRootUrl}templates/${this.currentTemplateId}/tabs/?group=experiment-navigation`;
```
This should use the template service

buehlefs commented on .gitignore:
```diff
@@ -25,7 +25,7 @@ chrome-profiler-events*.json
 # IDE - VSCode
 .vscode/*
-#!.vscode/settings.json
+!.vscode/settings.json
```
This will prevent users from having their own workspace specific settings.

buehlefs reviewed (commented):
These comments are for us to track what needs to be done before merging the PR. No need for you to change anything.
